go/merge/three_way.go

```diff
@@ -0,0 +1,259 @@
+//  - If we are dealing with a struct:
+//    - If the same field is both removed and inserted wrt parent: conflict
+//    - If the same field is inserted wrt parent, but with different values: conflict
+//  - If we are dealing with a list:
+//    - If the same index is both removed and inserted wrt parent: conflict
+//    - If the same index is inserted wrt parent, but with different values: conflict
+//  - If we are dealing with a set:
+//    - If the same object is both removed and inserted wrt parent: conflict
+//
+// All other modifications are allowed.
+// Currently, ThreeWay() only works on types.Map.
+func ThreeWay(a, b, parent types.Value, vwr types.ValueReadWriter) (merged types.Value, err error) {
+	if a == nil && b == nil {
+		return parent, nil
+	} else if a == nil {
+		return parent, newMergeConflict("Cannot merge nil Value with %s.", b.Type().Describe())
+	} else if b == nil {
+		return parent, newMergeConflict("Cannot merge %s with nil value.", a.Type().Describe())
+	} else if unmergeable(a, b) {
+		return parent, newMergeConflict("Cannot merge %s with %s.", a.Type().Describe(), b.Type().Describe())
+	}
+
+	return threeWayMerge(a, b, parent, vwr)
+}
+
+// a and b cannot be merged if they are of different NomsKind, or if at least one of the two is nil, or if either is a Noms primitive.
+func unmergeable(a, b types.Value) bool {
+	if a != nil && b != nil {
+		aKind, bKind := a.Type().Kind(), b.Type().Kind()
+		return aKind != bKind || types.IsPrimitiveKind(aKind) || types.IsPrimitiveKind(bKind)
+	}
+	return true
+}
+
+func threeWayMerge(a, b, parent types.Value, vwr types.ValueReadWriter) (merged types.Value, err error) {
+	d.PanicIfTrue(a == nil || b == nil, "Merge candidates cannont be nil: a = %v, b = %v", a, b)
+	newTypeConflict := func() *ErrMergeConflict {
+		pDescription := "<nil>"
+		if parent != nil {
+			pDescription = parent.Type().Describe()
+		}
+		return newMergeConflict("Cannot merge %s and %s on top of %s.", a.Type().Describe(), b.Type().Describe(), pDescription)
+	}
+
+	switch a.Type().Kind() {
+	default:
+		return parent, newMergeConflict("Cannot merge %s.", a.Type().Describe())
+
+	case types.ListKind:
+		// TODO: Come up with a plan for List (BUG 148)
+		return parent, newMergeConflict("Cannot merge %s.", a.Type().Describe())
+
+	case types.MapKind:
+		if aMap, bMap, pMap, ok := mapAssert(a, b, parent); ok {
+			return threeWayMapMerge(aMap, bMap, pMap, vwr)
+		}
+
+	case types.RefKind:
+		if aValue, bValue, pValue, ok := refAssert(a, b, parent, vwr); ok {
+			merged, err := threeWayMerge(aValue, bValue, pValue, vwr)
+			if err != nil {
+				return parent, err
+			}
+			return vwr.WriteValue(merged), nil
+		}
+
+	case types.SetKind:
+		// TODO: Implement plan from BUG148
+		return parent, newMergeConflict("Cannot merge %s.", a.Type().Describe())
+
+	case types.StructKind:
+		if aStruct, bStruct, pStruct, ok := structAssert(a, b, parent); ok {
+			return threeWayStructMerge(aStruct, bStruct, pStruct, vwr)
```

Comment:
Yeah. I was considering making Struct work in this patch as well, then stuff started to get too big. I'd like to keep the struct code in, though, since the next patch I do would just reintroduce it anyway :-)